clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordBatchTest.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients.producer.internals;
+
+import java.nio.ByteBuffer;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.record.CompressionType;
+import org.apache.kafka.common.record.MemoryRecordsBuilder;
+import org.apache.kafka.common.record.Record;
+import org.apache.kafka.common.record.TimestampType;
+import org.junit.Test;
+
+import static org.junit.Assert.assertFalse;
+
+public class RecordBatchTest {
+
+    /**
+     * A RecordBatch configured using very large linger value, in combination
+     * with a timestamp preceding the create time of a RecordBatch, is
+     * interpreted correctly as not expired when the linger time is larger than
+     * the difference between now and create time by RecordBatch#maybeExpire.
+     */
+    @Test
+    public void testLargeLingerOldNowExpire() {
```

Comment:
Can we also have some boundary case tests for the other branches in that method? Since it's not obvious that `now` can be before the other values, it's good to have test cases for that.